indexing-service/src/main/java/io/druid/indexing/worker/WorkerTaskMonitor.java

```diff
@@ -78,164 +85,354 @@ public WorkerTaskMonitor(
+          @Override
+          public void onSuccess(TaskStatus result)
+          {
+            notices.add(new StatusNotice(task, result));
+          }
 
-            try {
-              workerCuratorCoordinator.updateAnnouncement(TaskAnnouncement.create(task, taskStatus));
-              log.info(
-                  "Job's finished. Completed [%s] with status [%s]",
-                  task.getId(),
-                  taskStatus.getStatusCode()
-              );
-            }
-            catch (Exception e) {
-              log.makeAlert(e, "Failed to update task status")
-                 .addData("task", task.getId())
-                 .emit();
-            }
+          @Override
+          public void onFailure(Throwable t)
+          {
+            notices.add(new StatusNotice(task, TaskStatus.failure(task.getId())));
           }
         }
     );
   }
 
-  private boolean isTaskRunning(final String taskId)
+  @LifecycleStop
+  public void stop()
   {
-    for (final Task runningTask : running) {
-      if (runningTask.getId().equals(taskId)) {
-        return true;
+    synchronized (this) {
+      Preconditions.checkState(started, "not started");
+
+      try {
+        exec.shutdownNow();
+        pathChildrenCache.close();
+        taskRunner.stop();
+
+        started = false;
+        log.info("Stopped WorkerTaskMonitor.");
+      }
+      catch (Exception e) {
+        log.makeAlert(e, "Exception stopping WorkerTaskMonitor")
+           .emit();
       }
     }
+  }
+
+  private static class TaskDetails
+  {
+    private final Task task;
+    private final ListenableFuture<TaskStatus> future;
+    private final long startTime;
+    private TaskStatus status;
+    private TaskLocation location;
 
-    return false;
+    public TaskDetails(Task task, ListenableFuture<TaskStatus> future)
+    {
+      this.task = task;
+      this.future = future;
+      this.startTime = System.currentTimeMillis();
+      this.status = TaskStatus.running(task.getId());
+      this.location = TaskLocation.unknown();
+    }
   }
 
-  @LifecycleStop
-  public void stop()
+  private interface Notice
   {
-    try {
-      pathChildrenCache.close();
-      exec.shutdownNow();
-      taskRunner.stop();
-    }
-    catch (Exception e) {
-      log.makeAlert(e, "Exception stopping WorkerTaskMonitor")
-         .addData("exception", e.toString())
-         .emit();
+    String getTaskId();
+
+    void handle();
+  }
+
+  private class RunNotice implements Notice
+  {
+    private final Task task;
+
+    public RunNotice(Task task)
+    {
+      this.task = task;
+    }
+
+    @Override
+    public String getTaskId()
+    {
+      return task.getId();
+    }
+
+    @Override
+    public void handle()
+    {
+      if (running.containsKey(task.getId())) {
+        log.warn(
+            "Got run notice for task [%s] that I am already running...",
+            task.getId()
+        );
+        workerCuratorCoordinator.unannounceTask(task.getId());
```

Comment:
renamed the methods to be clearer.

`unannounceTask` -> `removeTaskRunZnode`
`updateAnnouncement` -> `updateTaskStatusAnnouncement`